mitosheet/src/streamlit/MitoMessagePasser.tsx

```diff
@@ -0,0 +1,52 @@
+import {
+    Streamlit,
+    StreamlitComponentBase,
+    withStreamlitConnection,
+} from "streamlit-component-lib"
+import React, { ReactNode } from "react"
+
+
+/**
+ * This component is used to pass messages from the Mito iframe to the running 
+ * streamlit backend. 
+ * 
+ * The reason it exists:
+ * - A streamlit component can only communicate with the backend with the setComponentValue
+ *   function.
+ * - When the setComponentValue function is called, the Python script re-runs from the top.
+ * - This means that when we render the component in the Python script, we have not yet 
+ *   received the data from the Mito iframe, as the component value is the return value
+ *   of the Python component function. 
+ * - To get around this, we use this component to listen for messages from the Mito iframe
+ *   and then pass them to the Python script using the setComponentValue function.
+ * - As this component is always rendered before the Mito iframe, we can be sure that
+ *   the Python script will have received the message from the Mito iframe before it 
+ *   renders the Mito iframe.
```

Comment:
So my understanding is this: 

1. Every time the user makes an edit to their data (or even something like loading the summary statistics) we send that message to the backend through the setComponentValue which is the only way that the Mito streamlit component can communicate with the backend. But, every time the setComponentValue function is called, it triggers the entire streamlit app to re-render from top to bottom. Which means that the Mito streamlit component will re-render before it gets back the response from the backend. 
2. To solve this issue, we created another component, the Mito Message Passer. When the Mito component needs to communicate with the backend, instead of directly calling the setComponentValue function, we tell the Mito Message Passer to call the setComponentValue instead. 

After that I'm a little bit confused. Primarily because: 
1. If the problem is that anytime the setComponentValue function is called, the entire streamlit app rerenders, then why does this approach fix that? We are still calling the setComponentValue function.

Reply:
Okay, I did some experimentation in a template streamlit component and I think this is my understanding to the above. 

When we call the setComponentValue function, the streamlit app starts pre-rendering right away, but doesn't finish rerendering until the component is finished updating. 

So I assume that what we're doing is that the MitoMessagePasser's value is actually the response of each message we pass. That way, because the MitoMessagePasser is rendered before the Mito Streamlit Component, we are sure to get the message back. 

The reason that this works is because we're actually setting the value of the component to the result, instead of just trying to get a response. 

I feel like that is on the right track, but not the full answer.

Reply:
With that understanding, it seems like another solution (not that we should do it, but just for my understanding) would be that instead of having API.tsx calls return responses, we send all of the responses through the components props, just like we do with SheetData, for example. 

That way, the component won't finish rerendering until it gets the data back. 

I know that there are other issues, like debouncing, and the fact that we don't actually want to rerender the entire app each time is possible (or at least make that configurable).

Reply:
When you call setComponentValue, the value appears in the Python script as the return value of the component function. The component function is also what takes the props. So we need the response before we call the component function, but we only get the message _after_ we call the component function (it's the return value, as I just said). 

The MitoMessagePasser lives before the Mito Component, and since we pass messages through this, it allows us to get messages before rending the Mito Component, and then pass the responses back with props.